README.md


![logo](https://raw.githubusercontent.com/Husseinhj/FSCalendar/master/ScreenShots/FSCalendar-Persian.png)
<br>
[![Build Status](https://travis-ci.org/Husseinhj/FSCalendar.svg?branch=master)](https://travis-ci.org/Husseinhj/FSCalendar)
[![Version](https://img.shields.io/cocoapods/v/FSCalendar+Persian.svg?style=flat)](http://cocoadocs.org/docsets/FSCalendar+Persian)

[![Platform](https://img.shields.io/badge/platform-iOS%207%2B-blue.svg?style=flat)](http://cocoadocs.org/docsets/FSCalendar+Persian)
[![Carthage compatible](https://img.shields.io/cocoapods/p/FSCalendar+Persian.svg)](https://github.com/Carthage/Carthage)
<br>
[![Languages](https://img.shields.io/badge/language-objc%20|%20swift-FF69B4.svg?style=plastic)](#)

# Table of contents
* [Screenshots](#screenshots)
* [Installation](#installation)
* [Pre-knowledge](#pre-knowledge)
* [Support](#support)
* [Contact](#contact)

## <a id="screenshots"></a>Screenshots 

### Persian Calendar and Arabic Calendar

| ![Week](https://github.com/Husseinhj/FSCalendar/raw/master/ScreenShots/Simulator%20Screen%20Shot%20-%20iPhone%208%20Plus%20-%202018-03-01%20at%2014.32.47.png) | ![Month](https://github.com/Husseinhj/FSCalendar/raw/master/ScreenShots/Simulator%20Screen%20Shot%20-%20iPhone%208%20Plus%20-%202018-03-01%20at%2014.33.06.png) | ![DIY](https://github.com/Husseinhj/FSCalendar/raw/master/ScreenShots/Simulator%20Screen%20Shot%20-%20iPhone%208%20Plus%20-%202018-03-01%20at%2016.46.44.png) |
| ---------------------------------------- | ---------------------------------------- | ---------------------------------------- |
| ![Range](https://github.com/Husseinhj/FSCalendar/raw/master/ScreenShots/Simulator%20Screen%20Shot%20-%20iPhone%208%20Plus%20-%202018-03-01%20at%2015.04.12.png) | ![Week range](https://github.com/Husseinhj/FSCalendar/raw/master/ScreenShots/Simulator%20Screen%20Shot%20-%20iPhone%208%20Plus%20-%202018-03-01%20at%2015.27.00.png) | ![Range selection](https://github.com/Husseinhj/FSCalendar/raw/master/ScreenShots/Simulator%20Screen%20Shot%20-%20iPhone%208%20Plus%20-%202018-03-01%20at%2015.26.50.png) |


### iPhone
![fscalendar](https://cloud.githubusercontent.com/assets/5186464/10262249/4fabae40-69f2-11e5-97ab-afbacd0a3da2.jpg)

### iPad
![fscalendar-ipad](https://cloud.githubusercontent.com/assets/5186464/10927681/d2448cb6-82dc-11e5-9d11-f664a06698a7.jpg)

### Safe Orientation

![fscalendar-scope-orientation-autolayout](https://cloud.githubusercontent.com/assets/5186464/20325758/ea125e1e-abc0-11e6-9e29-491acbcb0d07.gif)

### Today Extension
| iOS8/9                                   | iOS10                                    |
| ---------------------------------------- | ---------------------------------------- |
| ![today1](https://cloud.githubusercontent.com/assets/5186464/20288375/ed3fba0e-ab0d-11e6-8b15-43d3dc656f22.gif) | ![today2](https://cloud.githubusercontent.com/assets/5186464/20288378/f11e318c-ab0d-11e6-8d1d-9d89b563e9d7.gif) |

### Interactive Scope Gesture
| ![1](https://cloud.githubusercontent.com/assets/5186464/21559640/e92a9ccc-ce8a-11e6-8c60-e52204f33249.gif) |
| ---- |

### DIY support
|  ![1](https://cloud.githubusercontent.com/assets/5186464/20026983/22354a0e-a342-11e6-8ae6-0614ea7f35ae.gif)    |
| -------------  |
> To customize your own cell, view DIY Example in `Example-Swift` or `Example-Objc`


### Swipe-To-Choose

| Single-Selection<br/>Swipe-To-Choose     | Multiple-Selection<br/>Swipe-To-Choose   | DIY<br/>Swipe-To-Choose                  |
| ---------------------------------------- | ---------------------------------------- | ---------------------------------------- |
| ![1](https://cloud.githubusercontent.com/assets/5186464/20257768/cb1905d4-aa86-11e6-9ef7-af76f9caa024.gif) | ![2](https://cloud.githubusercontent.com/assets/5186464/20257826/254070ec-aa87-11e6-81b1-1815453fd464.gif) | ![3](https://cloud.githubusercontent.com/assets/5186464/20257836/2ffa3252-aa87-11e6-8ff9-3b40f5b2307b.gif) |

## Achievement of Users <a id="achievement"></a>

|  ![1](https://cloud.githubusercontent.com/assets/5186464/21747193/3111e4ee-d59a-11e6-8e4d-ca695b53e421.png)    |  ![2](https://cloud.githubusercontent.com/assets/5186464/21747393/42a753fa-d5a0-11e6-9cb2-de7cc642e69e.png)  |  ![3](https://cloud.githubusercontent.com/assets/5186464/21897255/ff78fcdc-d923-11e6-9d59-62119bc4343f.png)  |  ![4](https://cloud.githubusercontent.com/assets/5186464/21747192/3111cacc-d59a-11e6-8626-44cd75ebd794.png)  
|
| ------------- | ------------- | ------------- | ------------- |

#### [***More Achievements***](https://github.com/WenchaoD/FSCalendar/wiki/) are available in [***FSCalendar Gallery***](https://github.com/WenchaoD/FSCalendar/wiki/)  

# <a id="installation"></a>Installation

## CocoaPods of RTL support Version:

* For iOS8+: 👍

```ruby
use_frameworks!
target '<Your Target Name>' do
    pod 'FSCalendar+Persian'
end
```

* For iOS7+:

```ruby
target '<Your Target Name>' do
	pod 'FSCalendar+Persian'
end
```

> [NSCalendarExtension](https://github.com/WenchaoD/NSCalendarExtension) is required to get iOS7 compatibility.

## Carthage: 
* For iOS8+

```ruby
github "hussein_hj/FSCalendar"
```

## Manually:
* Drag all files under `FSCalendar` folder into your project. 👍

> Alternatively to give it a test run, simply press `command+u` in `Example-Objc` or `Example-Swift` and launch the ***UITest Target***. <br>
> Only the methods marked "👍" support IBInspectable / IBDesignable feature. [Have fun with Interface builder](#roll_with_interface_builder)


# Setup

## Use Interface Builder

1、 Drag an UIView object to ViewController Scene
2、 Change the `Custom Class` to `FSCalendar`<br/>
3、 Link `dataSource` and `delegate` to the ViewController <br/>

![fscalendar-ib](https://cloud.githubusercontent.com/assets/5186464/9488580/a360297e-4c0d-11e5-8548-ee9274e7c4af.jpg)

4、 Finally, implement `FSCalendarDataSource` and `FSCalendarDelegate` in your `ViewController`

## Or use code

```objc
@property (weak , nonatomic) FSCalendar *calendar;
```
```objc
// In loadView(Recommended) or viewDidLoad
FSCalendar *calendar = [[FSCalendar alloc] initWithFrame:CGRectMake(0, 0, 320, 300)];
calendar.dataSource = self;
calendar.delegate = self;
[self.view addSubview:calendar];
self.calendar = calendar;
```
<br/>

## Or swift

* To use `FSCalendar` in swift, you need to [Create Bridge Header](https://developer.apple.com/library/ios/documentation/Swift/Conceptual/BuildingCocoaApps/MixandMatch.html) first.


```swift
fileprivate weak var calendar: FSCalendar!
```
```swift
// In loadView or viewDidLoad
let calendar = FSCalendar(frame: CGRect(x: 0, y: 0, width: 320, height: 300))
calendar.dataSource = self
calendar.delegate = self
view.addSubview(calendar)
self.calendar = calendar
```

> To use **FSCalendar** in Swift3, see `Example-Swift` for details.

## How to use RTL calendar?

It's easy (Objective-C) :
```objc
calendar.locale = [NSLocale localeWithLocaleIdentifier:@"fa-IR"];
calendar.identifier = NSCalendarIdentifierPersian;
calendar.firstWeekday = 7;

....

#pragma mark - FSCalendarDataSource

- (NSDate *)minimumDateForCalendar:(FSCalendar *)calendar {
    return [self.dateFormatter dateFromString:@"2016-07-08"];
}
```

Swift :

```swift

calendar.locale = NSLocale.init(localeIdentifier: "fa-IR") as Locale
calendar.identifier = NSCalendar.Identifier.persian.rawValue
calendar.firstWeekday = 7

....

// MARK:- FSCalendarDataSource
    
func minimumDate(for calendar: FSCalendar) -> Date {
    return self.formatter.date(from: "2016-07-08")!
}

```
Also can see [here](https://github.com/Husseinhj/FSCalendar/commit/0ed037e852f4dc9b0d7569c00e21e75839aae7ef#diff-72ae6ec586cbfec0236fb013c01c9f20) where is these codes

## <a id='adjusts_frame_dynamicly' /></a>Warning 
`FSCalendar` ***doesn't*** update frame by itself, Please implement

* For ***AutoLayout***

```objc
- (void)calendar:(FSCalendar *)calendar boundingRectWillChange:(CGRect)bounds animated:(BOOL)animated
{
    self.calendarHeightConstraint.constant = CGRectGetHeight(bounds);
    // Do other updates here
    [self.view layoutIfNeeded];
}
```

* For ***Manual Layout***

```objc
- (void)calendar:(FSCalendar *)calendar boundingRectWillChange:(CGRect)bounds animated:(BOOL)animated
{
    calendar.frame = (CGRect){calendar.frame.origin,bounds.size};
    // Do other updates here
}
```

* If you are using ***Masonry***

```objc
- (void)calendar:(FSCalendar *)calendar boundingRectWillChange:(CGRect)bounds animated:(BOOL)animated
{
    [calendar mas_updateConstraints:^(MASConstraintMaker *make) {
        make.height.equalTo(@(bounds.size.height));
        // Do other updates
    }];
    [self.view layoutIfNeeded];
}
```

* If you are using ***SnapKit***

```swift
func calendar(_ calendar: FSCalendar, boundingRectWillChange bounds: CGRect, animated: Bool) {
    calendar.snp.updateConstraints { (make) in
        make.height.equalTo(bounds.height)
        // Do other updates
    }
    self.view.layoutIfNeeded()
}
```

### <a id="roll_with_interface_builder"></a> Roll with Interface Builder
![fscalendar - ibdesignable](https://cloud.githubusercontent.com/assets/5186464/9301716/2e76a2ca-4503-11e5-8450-1fa7aa93e9fd.gif)

# <a id="pre-knowledge"></a>Pre-knowledge

> In `Swift3`, `NSDate` and `NSDateFormatter` have been renamed to ***Date*** and ***DateFormatter*** , see `Example-Swift` for details.

## How to create NSDate object

* By **NSCalendar**.

```objc
self.gregorian = [NSCalendar calendarWithIdentifier:NSCalendarIdentifierGregorian];
```

Then:

```objc
NSDate *date = [gregorian dateWithEra:1 year:2016 month:9 day:10 hour:0 minute:0 second:0 nanosecond:0];
// 2016-09-10 00:00:00
```


* Or by **NSDateFormatter**

```objc
self.formatter = [[NSDateFormatter alloc] init];
self.formatter.dateFormat = @"yyyy-MM-dd";
```

Then:

```objc
NSDate *date = [self.formatter dateFromString:@"2016-09-10"];
```

## How to print out NSDate object

* Use **NSDateFormatter**

```objc
self.formatter = [[NSDateFormatter alloc] init];
self.formatter.dateFormat = @"yyyy/MM/dd";
```

```objc
NSString *string = [self.formatter stringFromDate:date];
NSLog(@"Date is %@", string);
```

## How to manipulate NSDate with NSCalendar

```objc
self.gregorian = [NSCalendar calendarWithIdentifier:NSCalendarIdentifierGregorian];
```
* Get component of NSDate

```objc
NSInteger era = [self.gregorian component:NSCalendarUnitEra fromDate:date];
NSInteger year = [self.gregorian component:NSCalendarUnitYear fromDate:date];
NSInteger month = [self.gregorian component:NSCalendarUnitMonth fromDate:date];
NSInteger day = [self.gregorian component:NSCalendarUnitDay fromDate:date];
NSInteger hour = [self.gregorian component:NSCalendarUnitHour fromDate:date];
NSInteger minute = [self.gregorian component:NSCalendarUnitMinute fromDate:date];
...

```

* Get next **month**

```objc
NSDate *nextMonth = [self.gregorain dateByAddingUnit:NSCalendarUnitMonth value:1 toDate:date options:0];
```

* Get next **day**

```objc
NSDate *nextDay = [self.gregorain dateByAddingUnit:NSCalendarUnitDay value:1 toDate:date options:0];
```

* Is date in today/tomorrow/yesterday/weekend

```objc
BOOL isToday = [self.gregorian isDateInToday:date];
BOOL isYesterday = [self.gregorian isDateInYesterday:date];
BOOL isTomorrow = [self.gregorian isDateInTomorrow:date];
BOOL isWeekend = [self.gregorian isDateInWeekend:date];
```

* Compare two dates

```objc

BOOL sameDay = [self.gregorian isDate:date1 inSameDayAsDate:date2];
// Yes if the date1 and date2 are in same day


[self.gregorian compareDate:date1 toDate:date2 toUnitGranularity:unit];
// compare the era/year/month/day/hour/minute .etc ...
// return NSOrderAscending/NSOrderSame/NSOrderDecending

BOOL inSameUnit = [self.gregorian isDate:date1 equalToDate:date2 toUnitGranularity:unit];
// if the given unit (era/year/month/day/hour/minute .etc) are the same


```

## <a id='contact'/></a> Contact
* Telegram： [**@Hussein_hj**](https://t.me/Hussein_hj)
* Twitter：[**@Hussein_Juybari**](https://twitter.com/hussein_juybari)

> If your made a beautiful calendar with this library in your app, please take a screen shot and [@me](https://twitter.com/hussein_juybari) in twitter. Your help really means a lot to me! <br/>


# License
FSCalendar is available under the MIT license. See the LICENSE file for more info.

### [Documentation](http://cocoadocs.org/docsets/FSCalendar/) | [More Usage](https://github.com/WenchaoD/FSCalendar/blob/master/MOREUSAGE.md) | 


CHANGELOG.md
# Change Log

## [2.9.2](https://github.com/Husseinhj/FSCalendar/releases/tag/2.9.2)
### Changes :
- Fix bug crash library in Persian locale if `minimumDate` method wasn't implemented by developer. now implementing the `minimumDate()` not force anymore.

## [2.9.1](https://github.com/Husseinhj/FSCalendar/releases/tag/2.9.1)
### Changes :
- Fix bug for supporting iOS 7.0

## [2.9.0](https://github.com/Husseinhj/FSCalendar/releases/tag/2.9.0)
### Changes :
- Fix bug show weeks title.
- Fix bug show current day wrong
- Fix bug show current month

### Upgrade note:
- Change `identifier` to `calendarIdentifier` property.

## [2.8.1](https://github.com/Husseinhj/FSCalendar/releases/tag/2.8.1)
- Fix reload bug

## [2.8.0](https://github.com/Husseinhj/FSCalendar/releases/tag/2.8.0)
-   Supporting RTL calendars like Persian and Arabic

## [2.7.4](https://github.com/WenchaoD/FSCalendar/tree/2.7.4) (2017-03-15)

[Full Changelog](https://github.com/WenchaoD/FSCalendar/compare/2.7.3...2.7.4)



****Closed issues:****



- 自定义cell，快速滚动页面时候，可能出现崩溃 [\#621](https://github.com/WenchaoD/FSCalendar/issues/621)

- Incorrect NSDate returned [\#613](https://github.com/WenchaoD/FSCalendar/issues/613)

- Pan Gesture causing crashing [\#612](https://github.com/WenchaoD/FSCalendar/issues/612)

- 未检查代理是否实现了协议方法就调用，导致找不到方法 [\#611](https://github.com/WenchaoD/FSCalendar/issues/611)

- Is it possible to deselect the date in FScalendar other  than user's deselection? [\#610](https://github.com/WenchaoD/FSCalendar/issues/610)

- is it possible to select dates in the range in recent version of fscalender? [\#609](https://github.com/WenchaoD/FSCalendar/issues/609)

- CFBundleExecutable key [\#607](https://github.com/WenchaoD/FSCalendar/issues/607)

- FSCalendar crash most of the time when minimumDate & maximumDate is set Today and selectDate is also Today [\#604](https://github.com/WenchaoD/FSCalendar/issues/604)

- a [\#603](https://github.com/WenchaoD/FSCalendar/issues/603)

- Today date does not update after changing device date [\#600](https://github.com/WenchaoD/FSCalendar/issues/600)

- 外观修改 [\#599](https://github.com/WenchaoD/FSCalendar/issues/599)

- Reloading of calendar view crashing app. [\#597](https://github.com/WenchaoD/FSCalendar/issues/597)



## [2.7.3](https://github.com/WenchaoD/FSCalendar/tree/2.7.3) (2017-03-03)

[Full Changelog](https://github.com/WenchaoD/FSCalendar/compare/2.7.2...2.7.3)



****Closed issues:****



- is there a faster way if i want to go to a specific year [\#591](https://github.com/WenchaoD/FSCalendar/issues/591)

- Reload issue in FSCalendar [\#587](https://github.com/WenchaoD/FSCalendar/issues/587)

- Calendar crashes when minimumDate & maximumDate is set [\#575](https://github.com/WenchaoD/FSCalendar/issues/575)

- disable some date [\#561](https://github.com/WenchaoD/FSCalendar/issues/561)

- SwiftExample app: First row of days starts with the last week of the last month in some cases [\#548](https://github.com/WenchaoD/FSCalendar/issues/548)

- 预给了日期范围后，重值日期范围有问题 [\#545](https://github.com/WenchaoD/FSCalendar/issues/545)

- 您好大神，问您一下关于headerTitle的问题 [\#536](https://github.com/WenchaoD/FSCalendar/issues/536)

- Accessibility / VoiceOver reading order should be left to right \(or numeric order\), not top to bottom [\#466](https://github.com/WenchaoD/FSCalendar/issues/466)



****Merged pull requests:****



- Make selectedDate nullable [\#581](https://github.com/WenchaoD/FSCalendar/pull/581) ([roark31337](https://github.com/roark31337))



## [2.7.2](https://github.com/WenchaoD/FSCalendar/tree/2.7.2) (2017-02-16)

[Full Changelog](https://github.com/WenchaoD/FSCalendar/compare/2.7.1...2.7.2)



****Closed issues:****



- App crashing when implement delegate methods [\#590](https://github.com/WenchaoD/FSCalendar/issues/590)

- Calendar only showing Sundays [\#589](https://github.com/WenchaoD/FSCalendar/issues/589)

- 你好,请问下如何自定义Header? [\#588](https://github.com/WenchaoD/FSCalendar/issues/588)

- Calendar not reloading min/max dates when calling reload data.  [\#585](https://github.com/WenchaoD/FSCalendar/issues/585)

- Incorrect date gets selected on using the default select method [\#580](https://github.com/WenchaoD/FSCalendar/issues/580)

- 日历底部线条怎么隐藏？ [\#579](https://github.com/WenchaoD/FSCalendar/issues/579)

- Weird crash [\#574](https://github.com/WenchaoD/FSCalendar/issues/574)

- Scroll direction changes from horizontal to vertical when changing from week view to month view using gestures [\#573](https://github.com/WenchaoD/FSCalendar/issues/573)

- Use of undeclared type 'FSCalendarMonthPosition'  [\#568](https://github.com/WenchaoD/FSCalendar/issues/568)

- 重用性 [\#562](https://github.com/WenchaoD/FSCalendar/issues/562)

- 你好！我怎么修改头部的日期显示的格式呢？另外我怎么修改底部线条的隐藏与否呢？ [\#556](https://github.com/WenchaoD/FSCalendar/issues/556)

- Version 2.7.0 is bugged [\#549](https://github.com/WenchaoD/FSCalendar/issues/549)

- Interface BuilderDesignables build failed [\#543](https://github.com/WenchaoD/FSCalendar/issues/543)

- error: IB Designables: Failed to update auto layout status: The agent crashed [\#538](https://github.com/WenchaoD/FSCalendar/issues/538)

- Any way to increase the size of event indicator? [\#525](https://github.com/WenchaoD/FSCalendar/issues/525)

- The week scope is not reset when returning to week view from a partial drag [\#473](https://github.com/WenchaoD/FSCalendar/issues/473)



## [2.7.1](https://github.com/WenchaoD/FSCalendar/tree/2.7.1) (2017-02-02)

[Full Changelog](https://github.com/WenchaoD/FSCalendar/compare/2.7.0...2.7.1)



****Closed issues:****



- Calendar showing previous month for current page, but current page in stack has the correct month. [\#566](https://github.com/WenchaoD/FSCalendar/issues/566)

- Storyboard not update interface when change Attributes Inspector [\#564](https://github.com/WenchaoD/FSCalendar/issues/564)


Example-Objc/StoryboardExampleViewController.m
//
//  StoryboardExampleViewController.m
//  Chinese-Lunar-Calendar
//
//  Created by Wenchao Ding on 01/29/2015.
//  Copyright (c) 2014 Wenchao Ding. All rights reserved.
//

#import "StoryboardExampleViewController.h"
#import "LunarFormatter.h"
#import "CalendarConfigViewController.h"

@interface StoryboardExampleViewController()<FSCalendarDataSource,FSCalendarDelegate,FSCalendarDelegateAppearance>

@property (weak  , nonatomic) IBOutlet FSCalendar *calendar;
@property (weak  , nonatomic) IBOutlet NSLayoutConstraint *calendarHeightConstraint;

@property (assign, nonatomic) NSInteger      theme;
@property (assign, nonatomic) BOOL           lunar;

@property (strong, nonatomic) NSArray<NSString *> *datesShouldNotBeSelected;
@property (strong, nonatomic) NSArray<NSString *> *datesWithEvent;

@property (strong, nonatomic) NSCalendar *gregorianCalendar;

@property (strong, nonatomic) NSDateFormatter *dateFormatter1;
@property (strong, nonatomic) NSDateFormatter *dateFormatter2;
@property (strong, nonatomic) LunarFormatter *lunarFormatter;

- (IBAction)unwind2StoryboardExample:(UIStoryboardSegue *)segue;

@end

@implementation StoryboardExampleViewController

#pragma mark - Life Cycle

- (instancetype)initWithCoder:(NSCoder *)coder
{
    self = [super initWithCoder:coder];
    if (self) {
        
        self.gregorianCalendar = [[NSCalendar alloc] initWithCalendarIdentifier:NSCalendarIdentifierPersian];
        
        NSLocale *persianLocale = [NSLocale localeWithLocaleIdentifier:@"fa-IR"];
        
        self.dateFormatter1 = [[NSDateFormatter alloc] init];
        self.dateFormatter1.locale = persianLocale;
        self.dateFormatter1.dateFormat = @"yyyy/MM/dd";

        self.dateFormatter2 = [[NSDateFormatter alloc] init];
        self.dateFormatter2.locale = persianLocale;
        self.dateFormatter2.dateFormat = @"yyyy-MM-dd";
        
        self.calendar.appearance.caseOptions = FSCalendarCaseOptionsHeaderUsesUpperCase|FSCalendarCaseOptionsWeekdayUsesUpperCase;
        
        self.datesShouldNotBeSelected = @[@"2016/08/07",
                                          @"2016/09/07",
                                          @"2016/10/07",
                                          @"2016/11/07",
                                          @"2016/12/07",
                                          @"2016/01/07",
                                          @"2016/02/07"];
        
        self.datesWithEvent = @[@"2016-12-03",
                                @"2016-12-07",
                                @"2016-12-15",
                                @"2016-12-25"];
        
        self.lunarFormatter = [[LunarFormatter alloc] init];
    }
    return self;
}

- (void)viewDidLoad
{
    [super viewDidLoad];
    
    self.navigationItem.backBarButtonItem = [[UIBarButtonItem alloc] initWithTitle:@"" style:UIBarButtonItemStylePlain target:nil action:nil];
    
    if ([[UIDevice currentDevice].model hasPrefix:@"iPad"]) {
        self.calendarHeightConstraint.constant = 400;
    }
    self.calendar.identifier = NSCalendarIdentifierPersian;
    self.calendar.locale = [NSLocale localeWithLocaleIdentifier:@"fa-IR"];
    self.calendar.firstWeekday = 7;
    
    self.calendar.accessibilityIdentifier = @"calendar";
    [self.calendar selectDate:[NSDate date] scrollToDate:YES];

}

- (void)dealloc
{
    NSLog(@"%s",__FUNCTION__);
}

#pragma mark - FSCalendarDataSource

- (NSString *)calendar:(FSCalendar *)calendar titleForDate:(NSDate *)date
{
    return [self.gregorianCalendar isDateInToday:date] ? @"0" : nil;
}

- (NSString *)calendar:(FSCalendar *)calendar subtitleForDate:(NSDate *)date
{
    if (!_lunar) {
        return nil;
    }
    return [self.lunarFormatter stringFromDate:date];
}

- (NSInteger)calendar:(FSCalendar *)calendar numberOfEventsForDate:(NSDate *)date
{
    if ([self.datesWithEvent containsObject:[self.dateFormatter2 stringFromDate:date]]) {
        return 1;
    }
    return 0;
}

- (NSDate *)minimumDateForCalendar:(FSCalendar *)calendar
{
    return [self.dateFormatter1 dateFromString:@"1390/10/01"];
}

- (NSDate *)maximumDateForCalendar:(FSCalendar *)calendar
{
    return [self.dateFormatter1 dateFromString:@"1400/05/31"];
}

#pragma mark - FSCalendarDelegate

- (BOOL)calendar:(FSCalendar *)calendar shouldSelectDate:(NSDate *)date atMonthPosition:(FSCalendarMonthPosition)monthPosition
{
    BOOL shouldSelect = ![_datesShouldNotBeSelected containsObject:[self.dateFormatter1 stringFromDate:date]];
    if (!shouldSelect) {
        UIAlertController *alertController = [UIAlertController alertControllerWithTitle:@"FSCalendar" message:[NSString stringWithFormat:@"FSCalendar delegate forbid %@  to be selected",[self.dateFormatter1 stringFromDate:date]] preferredStyle:UIAlertControllerStyleAlert];
        [alertController addAction:[UIAlertAction actionWithTitle:@"OK" style:UIAlertActionStyleDefault handler:nil]];
        [self presentViewController:alertController animated:YES completion:nil];
    } else {
        NSLog(@"Should select date %@",[self.dateFormatter1 stringFromDate:date]);
    }
    return shouldSelect;
}

- (void)calendar:(FSCalendar *)calendar didSelectDate:(NSDate *)date atMonthPosition:(FSCalendarMonthPosition)monthPosition
{
    NSLog(@"did select date %@",[self.dateFormatter1 stringFromDate:date]);
    if (monthPosition == FSCalendarMonthPositionNext || monthPosition == FSCalendarMonthPositionPrevious) {
        [calendar setCurrentPage:date animated:YES];
    }
}

- (void)calendarCurrentPageDidChange:(FSCalendar *)calendar
{
    NSLog(@"did change to page %@",[self.dateFormatter1 stringFromDate:calendar.currentPage]);
}

- (void)calendar:(FSCalendar *)calendar boundingRectWillChange:(CGRect)bounds animated:(BOOL)animated
{
    _calendarHeightConstraint.constant = CGRectGetHeight(bounds);
    [self.view layoutIfNeeded];
}

- (CGPoint)calendar:(FSCalendar *)calendar appearance:(FSCalendarAppearance *)appearance titleOffsetForDate:(NSDate *)date
{
    if ([self calendar:calendar subtitleForDate:date]) {
        return CGPointZero;
    }
    if ([_datesWithEvent containsObject:[self.dateFormatter2 stringFromDate:date]]) {
        return CGPointMake(0, -2);
    }
    return CGPointZero;
}

- (CGPoint)calendar:(FSCalendar *)calendar appearance:(FSCalendarAppearance *)appearance eventOffsetForDate:(NSDate *)date
{
    if ([self calendar:calendar subtitleForDate:date]) {
        return CGPointZero;
    }
    if ([_datesWithEvent containsObject:[self.dateFormatter2 stringFromDate:date]]) {
        return CGPointMake(0, -10);
    }
    return CGPointZero;
}

- (NSArray<UIColor *> *)calendar:(FSCalendar *)calendar appearance:(FSCalendarAppearance *)appearance eventSelectionColorsForDate:(nonnull NSDate *)date
{
    if ([self calendar:calendar subtitleForDate:date]) {
        return @[appearance.eventDefaultColor];
    }
    if ([_datesWithEvent containsObject:[self.dateFormatter2 stringFromDate:date]]) {
        return @[[UIColor whiteColor]];
    }
    return nil;
}

#pragma mark - Navigation

- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender
{
    CalendarConfigViewController *config = segue.destinationViewController;
    config.lunar = self.lunar;
    config.theme = self.theme;
    config.selectedDate = self.calendar.selectedDate;
    config.firstWeekday = self.calendar.firstWeekday;
    config.scrollDirection = self.calendar.scrollDirection;
}

- (void)unwind2StoryboardExample:(UIStoryboardSegue *)segue
{
    CalendarConfigViewController *config = segue.sourceViewController;
    self.lunar = config.lunar;
    self.theme = config.theme;
    [self.calendar selectDate:config.selectedDate scrollToDate:NO];
    
    if (self.calendar.firstWeekday != config.firstWeekday) {
        self.calendar.firstWeekday = config.firstWeekday;
    }
    
    if (self.calendar.scrollDirection != config.scrollDirection) {
        self.calendar.scrollDirection = config.scrollDirection;
        UIAlertController *alertController = [UIAlertController alertControllerWithTitle:@"FSCalendar" message:[NSString stringWithFormat:@"Now swipe %@",@[@"Vertically", @"Horizontally"][self.calendar.scrollDirection]] preferredStyle:UIAlertControllerStyleAlert];
        [alertController addAction:[UIAlertAction actionWithTitle:@"OK" style:UIAlertActionStyleDefault handler:nil]];
        [self presentViewController:alertController animated:YES completion:nil];
    }
}


#pragma mark - Private properties

- (void)setTheme:(NSInteger)theme
{
    if (_theme != theme) {
        _theme = theme;
        switch (theme) {
            case 0: {
                _calendar.appearance.weekdayTextColor = FSCalendarStandardTitleTextColor;
                _calendar.appearance.headerTitleColor = FSCalendarStandardTitleTextColor;
                _calendar.appearance.eventDefaultColor = FSCalendarStandardEventDotColor;
                _calendar.appearance.selectionColor = FSCalendarStandardSelectionColor;
                _calendar.appearance.headerDateFormat = @"MMMM yyyy";
                _calendar.appearance.todayColor = FSCalendarStandardTodayColor;
                _calendar.appearance.borderRadius = 1.0;
                _calendar.appearance.headerMinimumDissolvedAlpha = 0.2;
                break;
            }
            case 1: {
                _calendar.appearance.weekdayTextColor = [UIColor redColor];
                _calendar.appearance.headerTitleColor = [UIColor darkGrayColor];
                _calendar.appearance.eventDefaultColor = [UIColor greenColor];
                _calendar.appearance.selectionColor = [UIColor blueColor];
                _calendar.appearance.headerDateFormat = @"yyyy-MM";
                _calendar.appearance.todayColor = [UIColor redColor];
                _calendar.appearance.borderRadius = 1.0;
                _calendar.appearance.headerMinimumDissolvedAlpha = 0.0;
                
                break;
            }
            case 2: {
                _calendar.appearance.weekdayTextColor = [UIColor redColor];
                _calendar.appearance.headerTitleColor = [UIColor redColor];
                _calendar.appearance.eventDefaultColor = [UIColor greenColor];
                _calendar.appearance.selectionColor = [UIColor blueColor];
                _calendar.appearance.headerDateFormat = @"yyyy/MM";
                _calendar.appearance.todayColor = [UIColor orangeColor];
                _calendar.appearance.borderRadius = 0;
                _calendar.appearance.headerMinimumDissolvedAlpha = 1.0;
                break;
            }
            default:
                break;
        }
        
    }
}

- (void)setLunar:(BOOL)lunar
{
    if (_lunar != lunar) {
        _lunar = lunar;
        [_calendar reloadData];
    }
}

@end

